pkg/table/column.go

```diff
@@ -327,14 +324,21 @@ func handleZeroDatetime(sessVars *variable.SessionVars, col *model.ColumnInfo, c
 // Set it to true only in FillVirtualColumnValue and UnionScanExec.Next()
 // If the handle of err is changed latter, the behavior of forceIgnoreTruncate also need to change.
 // TODO: change the third arg to TypeField. Not pass ColumnInfo.
-func CastValue(sctx sessionctx.Context, val types.Datum, col *model.ColumnInfo, returnErr, forceIgnoreTruncate bool) (casted types.Datum, err error) {
-	return CastColumnValue(sctx.GetSessionVars(), val, col, returnErr, forceIgnoreTruncate)
+func CastValue(sctx variable.SessionVarsProvider, val types.Datum, col *model.ColumnInfo, returnErr, forceIgnoreTruncate bool) (casted types.Datum, err error) {
+	vars := sctx.GetSessionVars()
+	sc := vars.StmtCtx
+	return castColumnValue(sc.TypeCtx(), sc.ErrCtx(), vars.SQLMode, val, col, vars.ConnectionID, sc.InInsertStmt || sc.InUpdateStmt, returnErr, forceIgnoreTruncate)
 }
 
-// CastColumnValue casts a value based on column type.
-func CastColumnValue(vars *variable.SessionVars, val types.Datum, col *model.ColumnInfo, returnErr, forceIgnoreTruncate bool) (casted types.Datum, err error) {
-	sc := vars.StmtCtx
-	casted, err = val.ConvertTo(sc.TypeCtx(), &col.FieldType)
+// CastColumnValue casts a value based on column type with expression BuildContext
+func CastColumnValue(ctx expression.BuildContext, val types.Datum, col *model.ColumnInfo, returnErr, forceIgnoreTruncate bool) (casted types.Datum, err error) {
+	evalCtx := ctx.GetEvalCtx()
+	return castColumnValue(evalCtx.TypeCtx(), evalCtx.ErrCtx(), evalCtx.SQLMode(), val, col, ctx.ConnectionID(), ctx.InInsertOrUpdate(), returnErr, forceIgnoreTruncate)
```

Comment:
This function is only called here. Could we change it to `sc.InInsertStmt || sc.InUpdateStmt` like line330?

Reply:
We are going to remove `GetSessionVars` in  `BuildContext` to avoid couple too many fields with `StatementContext` and `SessionVars`. So there is no `InInsertStmt` and `InUpdateStmt` in it.